torch/nn/parallel/distributed.py

```diff
@@ -176,6 +187,20 @@ def train(self, mode=True):
         for module in self._module_copies[1:]:
             module.train(mode)
 
+    def recreate_groups(self):
```

Comment:
Why do you need to recreate groups?

Reply:
I'd rather have us understand all the quirks of NCCL2 before we merge this.

Reply:
Agreeing on understanding this is important. But this shouldn't be the blocker for this PR.  We currently don't have any "working" distributed training solution for PyTorch using Infiniband (The known deadlock). And this workaround at least "works" for now and I have trained imagenet for 120 Epochs on 8 nodes without any issues.

Reply:
It "works", but it makes the code of DDP and its usage much more complicated, which means trouble for all its current users and maintainers. I'm fine with prolonging changes to backends, but this is user visible part of the code, and it's already quite complicated, so I don't want to add more unless there's a good argument for it

Reply:
Ok, let's do this then. I am gonna remove part of these user-facing code in distributed.py. Let's only include the backend code and these new python functions in this PR instead. The distributed.py has already had the comment saying that it supports gloo only.